include/range/v3/view/all.hpp

```diff
@@ -15,86 +15,139 @@
 
 #include <type_traits>
 #include <meta/meta.hpp>
+#include <range/v3/begin_end.hpp>
+#include <range/v3/data.hpp>
+#include <range/v3/empty.hpp>
 #include <range/v3/range_fwd.hpp>
 #include <range/v3/range_concepts.hpp>
-#include <range/v3/iterator_range.hpp>
-#include <range/v3/begin_end.hpp>
 #include <range/v3/size.hpp>
+#include <range/v3/view_interface.hpp>
 #include <range/v3/utility/functional.hpp>
 #include <range/v3/utility/static_const.hpp>
 
 namespace ranges
 {
     inline namespace v3
     {
+        /// \cond
+        namespace detail
+        {
+            struct CanEmptyRange_
+              : concepts::refines<concepts::Range>
+            {
+                template<typename T>
+                auto requires_(T &t) -> decltype(
+                    concepts::valid_expr(
+                        empty(t)
+                    ));
+            };
+
+            template<typename Rng>
+            using CanEmptyRange = concepts::models<CanEmptyRange_, Rng>;
+        } // namespace detail
+        /// \endcond
+
         /// \addtogroup group-views
         /// @{
+        template<typename Rng>
+        struct ref_view
+          : view_interface<ref_view<Rng>, range_cardinality<Rng>::value>
+        {
+        private:
+            CONCEPT_ASSERT(std::is_object<Rng>::value && Range<Rng>() && !View<Rng>());
+
+            Rng *rng_ = nullptr;
+        public:
+            ref_view() = default;
+            constexpr explicit ref_view(Rng &rng) noexcept
+              : rng_{&rng}
+            {}
+            constexpr iterator_t<Rng> begin() const
+            RANGES_AUTO_RETURN_NOEXCEPT
+            (
+                ranges::begin(*rng_)
+            )
+            constexpr sentinel_t<Rng> end() const
+            RANGES_AUTO_RETURN_NOEXCEPT
+            (
+                ranges::end(*rng_)
+            )
+            template<typename R = Rng,
+                CONCEPT_REQUIRES_(SizedRange<R>())>
+            constexpr auto size() const
+                RANGES_DECLTYPE_NOEXCEPT(ranges::size(std::declval<R &>()))
+            {
+                return ranges::size(*rng_);
+            }
+            template<typename R = Rng,
+                CONCEPT_REQUIRES_(detail::CanEmptyRange<R>())>
+            constexpr auto empty() const
+                RANGES_DECLTYPE_NOEXCEPT(ranges::empty(std::declval<R &>()))
+            {
+                return ranges::empty(*rng_);
+            }
+            template<typename R = Rng,
+                CONCEPT_REQUIRES_(ContiguousRange<R>())>
+            constexpr auto data() const
+                RANGES_DECLTYPE_NOEXCEPT(ranges::data(std::declval<R &>()))
+            {
+                return ranges::data(*rng_);
+            }
+        };
+
+        template<typename R>
+        struct is_referenceable_range<ref_view<R>> : std::true_type {};
+
```

Comment:
Not a fan of this, as I describe [here](https://github.com/CaseyCarter/cmcstl2/pull/126#pullrequestreview-85859245).